web/packages/teleterm/src/services/tshd/createClient.ts

```diff
@@ -270,801 +176,475 @@ export function createTshdClient(
-                return;
-
-              case api.PasswordlessPrompt.CREDENTIAL:
-                const credResponse = index => {
-                  const credRes =
-                    api.LoginPasswordlessRequest_LoginPasswordlessCredentialResponse.create(
-                      { index }
-                    );
-                  stream.write(
-                    api.LoginPasswordlessRequest.create({
-                      request: {
-                        oneofKind: 'credential',
-                        credential: credRes,
-                      },
-                    })
-                  );
-                };
-
-                params.onPromptCallback({
-                  type: 'credential',
-                  onUserResponse: credResponse,
-                  data: { credentials: response.credentials || [] },
+              };
+
+              params.onPromptCallback({
+                type: 'pin',
+                onUserResponse: pinResponse,
+              });
+              return;
+
+            case api.PasswordlessPrompt.CREDENTIAL:
+              const credResponse = (index: number) => {
+                stream.requests.send({
+                  request: {
+                    oneofKind: 'credential',
+                    credential: { index },
+                  },
                 });
-                return;
-
-              case api.PasswordlessPrompt.TAP:
-                if (hasDeviceBeenTapped) {
-                  params.onPromptCallback({ type: 'retap' });
-                } else {
-                  hasDeviceBeenTapped = true;
-                  params.onPromptCallback({ type: 'tap' });
-                }
-                return;
-
-              // Following cases should never happen but just in case?
-              case api.PasswordlessPrompt.UNSPECIFIED:
-                stream.cancel();
-                return reject(
-                  new Error('no passwordless prompt was specified')
-                );
-
-              default:
-                stream.cancel();
-                return reject(
-                  new Error(
-                    `passwordless prompt '${response.prompt}' not supported`
-                  )
-                );
-            }
-          });
+              };
+
+              params.onPromptCallback({
+                type: 'credential',
+                onUserResponse: credResponse,
+                data: { credentials: response.credentials || [] },
+              });
+              return;
+
+            case api.PasswordlessPrompt.TAP:
+              if (hasDeviceBeenTapped) {
+                params.onPromptCallback({ type: 'retap' });
+              } else {
+                hasDeviceBeenTapped = true;
+                params.onPromptCallback({ type: 'tap' });
+              }
+              return;
+
+            // Following cases should never happen but just in case?
+            case api.PasswordlessPrompt.UNSPECIFIED:
+              stream.requests.complete();
```

Comment:
Is `complete` the same as `cancel`?

Reply:
Not really (but this doesn't seem to make a difference here).
`complete`: Complete / close the stream. Can only be called if there is no pending send(). No send() should follow this call.
`cancel`: Cancel the ongoing call. Results in the call ending with a CANCELLED status, unless it has already ended with some other status.

I think this doesn't matter here anyway, because just after calling these functions we do `return reject(new Error(..)`, so any gRPC response is discarded.
However, I feel like this code is (and was) prone to race conditions. There is no guarantee that `onMessage`/`onError` handler is called after our rejection.

Ideal solution for this would be to add `AbortController` inside the handler and call `abort` for the cases above. Then combine its `abortSignal` with the abort signal from the callsite with `AbortSignal.any`. 
Thanks to that, all errors would go the same route.
There two problems with this tho:
* `AbortSignal.any` is supported since Node.js 20.3 (but we update soon).
* The abort reason is not propagated, the only error we get is 'Cancelled on client' :( 

So I think we have to stay with `stream.requests.complete()`

Reply:
Oh I see what you mean about `AbortSignal.any`.

Is "Cancelled on client" generated by the library or by us? We could probably improve the UX a little bit by checking if the error is an abort error when displaying the error in the modal. But that's for another PR I suppose.

Reply:
Yeah, it'd mostly be the error message. "Cancelled on client" sounds much worse than not showing the error at all if we know that the login process was just cancelled by the user.

Reply:
We can think of it. For now, I will keep it as is, since we have this behavior from the beginning.